server/datastore/mysql/migrations/tables/20201011162341_CleanupSoftDeletedColumns.go

```diff
@@ -0,0 +1,95 @@
+package tables
+
+import (
+	"fmt"
+
+	"database/sql"
+	"github.com/kolide/goose"
+)
+
+func init() {
+	goose.AddMigration(Up20201011162341, Down20201011162341)
+}
+
+func cleanupSoftDeleteFields(tx *sql.Tx, dbTable string) error {
+	deleteStmt := fmt.Sprintf("DELETE FROM `%s` WHERE deleted;", dbTable)
+
+	_, err := tx.Exec(deleteStmt)
+	if err != nil {
+		return err
+	}
+
+	for _, column := range []string{"deleted", "deleted_at"} {
+		alterStmt := fmt.Sprint(
+			"ALTER TABLE `%s` DROP COLUMN `%s`;", dbTable, column)
```

Comment:
I want to point out a couple of concerns about running migrations like this, related to issues you and I have discussed recently:

1) Note that any DDL statement in MySQL (like an `ALTER`) will _implicitly_ end the current transaction. This is the source of the issues of the database ending up in an inconsistent state during migrations. Additionally, MySQL does not support an `IF EXISTS` clause for these statements. It may make sense to try to detect the specific error that could be encountered if the column does not exist and ignore it. In general, I'd like to look towards employing such strategies for DDL migrations.

2) If there are any Fleet servers of the prior version running after these `ALTER`s have executed, they will encounter many errors as they query for the nonexistent columns. I've considered the possibility of using a strategy in which we stop using columns in one release and actually drop them in the next. That would partially mitigate the problem, but additional work would have to be done to explicitly `SELECT` the desired columns so that mapping errors do not occur after new columns are added.

Both of the above problems are mitigated by our new documentation in https://github.com/kolide/fleet/pull/2279/files, but I am also considering what it would take to be able to improve availability during upgrades.

What do you think?

Reply:
(1) I'll investigate this and see if I can detect the specific case to ignore.

(2) I think this is technically covered by the recommendation to take fleet offline during migrations, but it's always nice to make migrations less painful / maintain higher availability.

I assume when you say 'stop using the columns' in one release, that would mean:
* Do a migration in the first release to drop all rows where `deleted` is true
* Stop using `deleted` fields in select / update statements in the first release, including removing Deleted fields from all the structs.
* In the **next** release, do the ALTER to drop the column.

I think this is a bit better, as it should allow the FIRST release version of fleet to continue operating without errors after the ALTER.

Let's say the releases are as follows:
* x.y.0 - last release before any changes
* x.y.1 - release that stops using delete columns
* x.y.2 - release that alters tables

If someone skips over x.y.1 release directly to x.y.2, they'll still hit this case, but I think that's unavoidable. I'd probably include that in the upgrading guide.

---

I'll look at (1) this weekend, but doing (2) will involve a lot more rework. Happy to take a look at starting that if you think it'd be worthwhile.

Reply:
Pushing a commit for (1), but it generally looks like:

```
_, err := tx.Exec(stmt)
if err != nil {
    mysqlErr, ok := err.(*mysql.MySQLError)
    if !ok {
        return err
    }

    // depending on whether this is alter or delete, compare against diff constant value
    if mysqlErr.Number != someConstant {
        return err
    }
    fmt.Println("some nice message about skipping because column didn't exist")
}
```

Reply:
In the new commit, the error messages will look like:
```
Skipped deleting 'soft-deleted' entries from '<table>' because the 'deleted' column does not exist.
Skipped dropping column 'deleted' on table '<table>' because column does not exist.
```

Reply:
Yeah I think you have everything pretty much spot-on there.

I think going ahead with (1) and addressing (2) in some later realease (if we decide to) makes sense.